core/service.go

```diff
@@ -89,32 +157,97 @@ func (srv *RelayService) Serve(ctx context.Context) error {
 
 	msgs := NewRelayMsgs()
 
+	// reset watch start time for packets
+	if len(pseqs.Src) > 0 {
+		resetWatchStartTime(&srv.optimizeRelay.srcRelayPacketStartTime)
+	}
+	if len(pseqs.Dst) > 0 {
+		resetWatchStartTime(&srv.optimizeRelay.dstRelayPacketStartTime)
+	}
+
+	// reset watch start time for acks
+	if len(aseqs.Src) > 0 {
+		resetWatchStartTime(&srv.optimizeRelay.srcRelayAckStartTime)
+	}
+	if len(aseqs.Dst) > 0 {
+		resetWatchStartTime(&srv.optimizeRelay.dstRelayAckStartTime)
+	}
+
+	doExecuteRelaySrc, doExecuteRelayDst := srv.shouldExecuteRelay(pseqs, srv.optimizeRelay.srcRelayPacketStartTime, srv.optimizeRelay.dstRelayPacketStartTime)
+	doExecuteAckSrc, doExecuteAckDst := srv.shouldExecuteRelay(aseqs, srv.optimizeRelay.srcRelayAckStartTime, srv.optimizeRelay.dstRelayAckStartTime)
+
 	// update clients
-	if m, err := srv.st.UpdateClients(srv.src, srv.dst, pseqs, aseqs, srv.sh, true); err != nil {
+	if m, err := srv.st.UpdateClients(srv.src, srv.dst, doExecuteRelaySrc, doExecuteRelayDst, doExecuteAckSrc, doExecuteAckDst, srv.sh, true); err != nil {
 		logger.Error("failed to update clients", err)
 		return err
 	} else {
 		msgs.Merge(m)
 	}
 
 	// relay packets if unrelayed seqs exist
-	if m, err := srv.st.RelayPackets(srv.src, srv.dst, pseqs, srv.sh); err != nil {
+	if m, err := srv.st.RelayPackets(srv.src, srv.dst, pseqs, srv.sh, doExecuteRelaySrc, doExecuteRelayDst); err != nil {
 		logger.Error("failed to relay packets", err)
 		return err
 	} else {
 		msgs.Merge(m)
 	}
 
+	if doExecuteRelaySrc {
+		srv.optimizeRelay.srcRelayPacketStartTime.AlreadySet = false
+	}
+	if doExecuteRelayDst {
+		srv.optimizeRelay.dstRelayPacketStartTime.AlreadySet = false
+	}
+
 	// relay acks if unrelayed seqs exist
-	if m, err := srv.st.RelayAcknowledgements(srv.src, srv.dst, aseqs, srv.sh); err != nil {
+	if m, err := srv.st.RelayAcknowledgements(srv.src, srv.dst, aseqs, srv.sh, doExecuteAckSrc, doExecuteAckDst); err != nil {
 		logger.Error("failed to relay acknowledgements", err)
 		return err
 	} else {
 		msgs.Merge(m)
 	}
+	if doExecuteAckSrc {
+		srv.optimizeRelay.srcRelayAckStartTime.AlreadySet = false
+	}
+	if doExecuteAckDst {
+		srv.optimizeRelay.dstRelayAckStartTime.AlreadySet = false
+	}
 
 	// send all msgs to src/dst chains
 	srv.st.Send(srv.src, srv.dst, msgs)
 
 	return nil
 }
+
+func (srv *RelayService) shouldExecuteRelay(seqs *RelayPackets, srcRelayStartTime, dstRelayStartTime WatchStartTime) (bool, bool) {
+	srcRelay := false
+	dstRelay := false
+
+	// packet count
+	srcRelayCount := len(seqs.Src)
+	dstRelayCount := len(seqs.Dst)
+	if int64(srcRelayCount) >= srv.optimizeRelay.srcOptimizeCount {
+		srcRelay = true
+	}
+	if int64(dstRelayCount) >= srv.optimizeRelay.dstOptimizeCount {
+		dstRelay = true
+	}
+
+	// time interval
+	if time.Since(srcRelayStartTime.StartTime) >= srv.optimizeRelay.srcOptimizeInterval {
```

Comment:
Don't you need check if `srcRelayStartTime.AlreadySet`?

Reply:
Add `srcRelayStartTime.AlreadySet`